### DISCUSSION TITLE
Callback added via dag_policy using cluster policy feature is not working

### DISCUSSION THREAD

**Discussion by Bowrna** [apache/airflow#28100] [General]
### Apache Airflow version

2.4.2

### What happened

I am using cluster policy feature to add callback for all the dag. So I have used dag_policy function and added the callback to be invoked for success and failure.

Case 1: Added failure callback via cluster policy and ran DAG(which fails). No callback is invoked

Case 2: Added failure callback locally in DAG and ran DAG. Callback for failure case is invoked.

Case 3: Added failure callback locally in DAG and via cluster policy too. Then ran the DAG. Now both local callback and callback added via cluster policy got executed.

### What you think should happen instead

When callback for DAG is added via cluster policy, it has to get executed.

### How to reproduce

This is the cluster policy I have added
https://gist.github.com/Bowrna/1994894beea39fa8e1c269591b7f0346

This is the test dag I have tried running
https://gist.github.com/Bowrna/7fdca8b546fc274edd068ffdae5b76f9

### Operating System

macOS Monterey 12.2.1

### Versions of Apache Airflow Providers

apache-airflow-providers-apache-hive==4.0.1
apache-airflow-providers-cncf-kubernetes==4.4.0
apache-airflow-providers-common-sql==1.2.0
apache-airflow-providers-ftp==3.1.0
apache-airflow-providers-http==4.0.0
apache-airflow-providers-imap==3.0.0
apache-airflow-providers-sqlite==3.2.1

### Deployment

Virtualenv installation

### Deployment details

I am using development setup and trying to use cluster policy with callback added

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by ephraimbuddy** [comment]
From the [doc, task/dag policies](https://airflow.apache.org/docs/apache-airflow/stable/concepts/cluster-policies.html) runs at load time.

```
dag_policy: Takes a DAG parameter called dag. Runs at load time.

task_policy: Takes a BaseOperator parameter called task. Runs at load time.
```

**Comment by Bowrna** [comment]
Even if it runs at load time, the callback param gets appended to the dag object. the callback has to be executed if the dag fails in that case. Am I missing to understand anything in this @ephraimbuddy?

**Comment by potiuk** [comment]
Airflow 2.5.0 (released three days ago) has currently much better solution for those kind of approach - listeners:

https://airflow.apache.org/docs/apache-airflow/stable/listeners.html

They are much better way of handling the lifecycle of DAGs and Tasks, and we strongly recommend to upgrade rather than try to add callbacks via policies. This is a much more stable and robust solution that will likely get improved and optimized in the future and it should be used instead of policies for any kind of "lifecycle" global callbacks. 

I will convert this one into a discussion now - in case more discussion is needed.

**Reply by Bowrna** [reply]
Yes, using listeners is a better solution than any of the other solutions. There is no doubt about that part.
But many users are still in the older version. Say example, Amazon MWAA supports Airflow version v2.2.2 as their latest version. That's why I am exploring this part for monitoring the DAG lifecycle. @potiuk

**Reply by potiuk** [reply]
But it's not going to work in this version (and we are not going to relase a change for that I am afraid). Better to focus time elsewhere.
